File: resources/calc_client.erl
-module(calc_client).
-author("Duncan Paul Attard").

-export([add/2, mul/2]).
-export([rpc/2]).

-spec add(A :: number(), B :: number()) -> number().
add(A, B) ->
    {add, Res} = rpc(calc_server, {add, A, B}),
    Res.

-spec mul(A :: number(), B :: number()) -> number().
mul(A, B) ->
    {mul, Res} = rpc(calc_server, {mul, A, B}),
    Res.

-spec rpc(To :: pid() | atom(), Req :: any()) -> any().
rpc(To, Req) ->
    To ! {self(), Req},
    receive
        Resp ->
            Resp
    end.

File: resources/calc_client.ex
defmodule Demo.CalcClient do
  @spec add(a :: number(), b :: number()) :: number()
  def add(a, b) do
    {:add, res} = rpc(Demo.CalcServer, {:add, a, b})
    res
  end

  @spec mul(a :: number(), b :: number()) :: number()
  def mul(a, b) do
    {:mul, res} = rpc(Demo.CalcServer, {:mul, a, b})
    res
  end

  @spec rpc(to :: pid() | atom(), req :: any()) :: any()
  def rpc(to, req) do
    send(to, {self(), req})

    receive do
      resp ->
        IO.inspect(resp)
        resp
    end
  end
end
